gitpod-io/gitpod
# [ws-proxy] Explicitly refresh state if we don't know about a workspace

This PR introduces an explicit state refresh in case of a workspace info cache miss.

### How does it work
We have a `refreshWorkspaceInfo` Go routine which at regular intervals polls ws-manager, if someone requested a refresh.
This way we don't place direct load on ws-manager if we don't have to. When such an explicit refresh happens, we'll log a warning.

### How to test
1. Attempt to access a workspace that doesn't exist, e.g. ...
  You should see a log message in ws-proxy: _refreshing info from ws-manager_
2. Run the tests often: `go test -v ./... -count 100`
